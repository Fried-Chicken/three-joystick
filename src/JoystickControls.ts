import {
  MeshLambertMaterial,
  CircleGeometry,
  Mesh,
  Scene,
  Vector2,
  Vector3,
  Quaternion,
  PerspectiveCamera,
  Object3D,
} from 'three';
import isTouchOutOfBounds from './helpers/isTouchOutOfBounds';

const degreesToRadians = (degrees: number): number => {
  return degrees * (Math.PI / 180);
};

class JoystickControls {
  scene: Scene;
  camera: PerspectiveCamera;
  joystickTouchZone = 75;
  environment: Object3D = new Object3D();
  quaternion: Quaternion = new Quaternion();
  xAxis = new Vector3(1, 0, 0);
  yAxis = new Vector3(0, 1, 0);
  rotateFactor = 0.0005;
  /**
   * Timestamp of when the user touched the screen.
   * This is used for de-bouncing the user interaction
   */
  touchStart = 0;
  /**
   * Anchor of the joystick base
   */
  baseAnchorPoint: Vector2 = new Vector2();
  /**
   * Current point of the joystick ball
   */
  ballAnchorPoint: Vector2 = new Vector2();
  /**
   * Function that allows you to prevent the joystick
   * from attaching
   */
  preventAction: () => boolean = () => false;
  /**
   * The minimum distance required for the joystick to
   * activate. Defaults to 50px
   */
  activateAfterPixelDistance = 50;
  /**
   * Creation of joystick triggered
   */
  isActive = false;
  /**
   * Creation of joystick complete
   */
  joystickIsActive = false;
  /**
   * Use mesh for now
   *
   * TODO: Make this outside the joystick
   */
  target: Object3D;

  constructor(
    camera: PerspectiveCamera,
    scene: Scene,
    target: Object3D,
  ) {
    this.camera = camera;
    this.scene = scene;
    this.target = target;
    this.createTouchEventListeners();
  }

  /**
   * This is used to determine whether a user has tapped/clicked or
   * swiped/dragged
   *
   * @param debounceInMs
   * @returns {boolean}
   */
  private debounceTime = (debounceInMs: number) => (
    (Date.now() - this.touchStart) < debounceInMs
  );

  public update = (): void => {
    this.createRotation();
  };

  /**
   * TODO: Find out why we don't do iPhone
   */
  private isAndroid = (event: TouchEvent) => {
    if (navigator.userAgent.match(/Android/i)) {
      event.preventDefault();
    }
  };

  private swipeDistanceIsMoreThan = (touch: TouchEvent) => {
    const distance = touch.touches?.item(0);

    if (distance === null) {
      return;
    }

    const xDistance = Math.abs(this.baseAnchorPoint.x - distance.clientX);
    const yDistance = Math.abs(this.baseAnchorPoint.y - distance.clientY);

    return (
      (xDistance > this.activateAfterPixelDistance) ||
      (yDistance > this.activateAfterPixelDistance)
    );
  };


  private createCircle = (
    name: string,
    position: Vector3,
    color: number,
    size: number,
  ) => {
    const geometry = new CircleGeometry(size, 72);
    const material = new MeshLambertMaterial({
      color: color,
      opacity: 0.5,
    });
    material.transparent = true;
    const joyStickBaseCircle = new Mesh(geometry, material);
    joyStickBaseCircle.name = name;
    joyStickBaseCircle.position.copy(position);
    this.scene.add(joyStickBaseCircle);
  };

  private updateJoystickBallPosition = (event: TouchEvent) => {
    const joyStickBase = this.scene.getObjectByName('joystick-base');
    const joyStickBall = this.scene.getObjectByName('joystick-ball');
    const zoomScale = 1 / this.camera.zoom;
    const touch = event.touches.item(0);

    if (touch === null) {
      return;
    }

    // TODO: Set min, max and hotzone once to this
    const touchX = (touch.clientX / window.innerWidth) * 2 - 1;
    const touchY = -(touch.clientY / window.innerHeight) * 2 + 1;
    const vector = new Vector3(touchX, touchY, 0.5).unproject(this.camera);
    let direction = vector.sub(this.camera.position).normalize();
    let position = this.camera.position.clone()
      .add(direction.multiplyScalar(10));

    if (joyStickBase && joyStickBall) {
      if (isTouchOutOfBounds(touch, this.baseAnchorPoint, this.joystickTouchZone)) {
        console.log('Outside Base');
        const angle = Math.atan2(
          touch.clientY - this.baseAnchorPoint.y,
          touch.clientX - this.baseAnchorPoint.x,
        ) - degreesToRadians(90);
        const xDistance = Math.sin(angle) * this.joystickTouchZone;
        const yDistance = Math.cos(angle) * this.joystickTouchZone;
        direction = new Vector3(-xDistance, -yDistance, 0).normalize();
        position = joyStickBase.position.clone().add(direction);
      } else {
        console.log('Inside Base');
      }
      joyStickBall.position.copy(position);
      return;
    }

    //** Joystick Base Position
    this.createCircle('joystick-base', position, 0x666666, 0.9 * zoomScale);
    this.createCircle('joystick-ball', position, 0x444444, 0.5 * zoomScale);
    this.joystickIsActive = true;
  };

  private removeJoystick = () => {
    const joystickBase = this.scene.getObjectByName('joystick-base');
    const joyStickBall = this.scene.getObjectByName('joystick-ball');

    if (joystickBase && joyStickBall) {
      this.scene.remove(joystickBase);
      this.scene.remove(joyStickBall);
    }

    this.joystickIsActive = false;
  };

  /**
   * Rotates the environment
   * @param axis
   * @param angle
   */
  private rotate = (axis: Vector3, angle: number) => {
    this.quaternion.setFromAxisAngle(axis, angle);
    this.target?.quaternion.premultiply(this.quaternion);
  };

  /**
   * createRotation function used to calculate the amount of rotation
   */
  private createRotation = (): void => {
    if (!this.isActive) {
      return;
    }

    /**
     * TODO: xAxis and yAxis must be set
     */
    const moveX = this.ballAnchorPoint.x - this.baseAnchorPoint.x;
    const moveY = this.ballAnchorPoint.y - this.baseAnchorPoint.y;

    this.rotate(this.yAxis, moveX * this.rotateFactor);
    this.rotate(this.xAxis, moveY * this.rotateFactor);
  };

  destroyTouchEventListeners = (): void => {
    /**
     * TODO: Remove event listeners from the document
     */
  };

  private createTouchEventListeners = () => {
    document.addEventListener('touchstart', (event: TouchEvent) => {
      if (this.preventAction()) {
        return;
      }

      const touch = event.touches.item(0);

      if (touch === null) {
        return;
      }

      this.baseAnchorPoint = new Vector2(touch.clientX, touch.clientY);
      this.touchStart = Date.now();
    });

    document.addEventListener('touchmove', (event: TouchEvent) => {
      if (this.debounceTime(150) && this.swipeDistanceIsMoreThan(event)) {
        // Return because we tapped instead
        return;
      }

      this.isActive = true;

      if (this.preventAction()) {
        return;
      }

      const touch = event.touches.item(0);

      if (!this.isActive) {
        return;
      }

      this.ballAnchorPoint = new Vector2(touch?.clientX, touch?.clientY);
      this.updateJoystickBallPosition(event);
    });

    document.addEventListener('touchend', () => {
      if (this.debounceTime(150) || !this.joystickIsActive) {
        return;
      }

      this.touchStart = 0;

      this.removeJoystick();

      setTimeout(() => {
        this.isActive = false;
      }, 150);
    });
  };
}

export default JoystickControls;
