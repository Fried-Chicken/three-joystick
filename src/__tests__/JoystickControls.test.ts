import * as THREE from 'three';
import JoystickControls from '../JoystickControls';
import { Object3D } from 'three';

// import getPositionInScene from '../helpers/getPositionInScene';

enum TOUCH {
  START = 'touchstart',
  MOVE = 'touchmove',
  END = 'touchend',
}

enum MOUSE {
  DOWN = 'mousedown',
  MOVE = 'mousemove',
  UP = 'mouseup',
}

const fireTouchEvent = (
  touchEventName: TOUCH,
  location?: { clientX: number, clientY: number },
) => {
  const touchEnd = new TouchEvent(touchEventName, {});
  touchEnd.touches.item = () => (location as Touch);
  window.dispatchEvent(touchEnd);
};

const fireMouseEvent = (
  mouseEventName: MOUSE,
  location?: { clientX: number, clientY: number },
) => {
  const mouseEvent = new MouseEvent(mouseEventName, location);

  window.dispatchEvent(mouseEvent);
};

jest.mock('../helpers/getPositionInScene', () => ({
  __esModule: true,
  default: () => new THREE.Vector2(100, 100),
}));

describe('JoystickControls', () => {
  describe('touch events', () => {
    const scene = new THREE.Scene();
    const camera = new THREE.PerspectiveCamera();

    it('should not invoke `onStart` if the `preventAction` function returned `true`', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );
      controls.preventAction = () => true;

      const spyOnStart = jest.spyOn(
        controls,
        // @ts-ignore
        'onStart',
      );

      fireTouchEvent(
        TOUCH.START,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      expect(spyOnStart).not.toHaveBeenCalled();
    });

    it('should not invoke `onStart` if the touch is undefined', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );
      const spyOnStart = jest.spyOn(
        controls,
        // @ts-ignore
        'onStart',
      );

      fireTouchEvent(TOUCH.START);

      expect(spyOnStart).not.toHaveBeenCalled();
    });

    it('should invoke `onStart` if the touch is defined', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );
      const spyOnStart = jest.spyOn(
        controls,
        // @ts-ignore
        'onStart',
      );

      fireTouchEvent(
        TOUCH.START,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      expect(spyOnStart).toHaveBeenCalledTimes(1);
    });

    it('should not invoke `onMove` if the `preventAction` function returned `true`', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );
      controls.preventAction = () => true;

      const spyOnMove = jest.spyOn(
        controls,
        // @ts-ignore
        'onMove',
      );

      fireTouchEvent(
        TOUCH.MOVE,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      expect(spyOnMove).not.toHaveBeenCalled();
    });

    it('should not invoke `onMove` if the touch is undefined', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );
      const spyOnMove = jest.spyOn(
        controls,
        // @ts-ignore
        'onMove',
      );

      fireTouchEvent(TOUCH.MOVE);

      expect(spyOnMove).not.toHaveBeenCalled();
    });

    it('should invoke `onMove` if the touch is defined', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );
      const spyOnMove = jest.spyOn(
        controls,
        // @ts-ignore
        'onMove',
      );

      fireTouchEvent(
        TOUCH.MOVE,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      expect(spyOnMove).toHaveBeenCalledTimes(1);
    });

    it('should not invoke `onEnd` if the joystick is not attached', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );
      const spyOnEnd = jest.spyOn(
        controls,
        // @ts-ignore
        'onEnd',
      );

      controls.isJoystickAttached = false;

      fireTouchEvent(TOUCH.END);

      expect(spyOnEnd).not.toHaveBeenCalled();
    });

    it('should invoke `onEnd` if the joystick is attached', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );
      const spyOnEnd = jest.spyOn(
        controls,
        // @ts-ignore
        'onEnd',
      );

      controls.isJoystickAttached = true;

      fireTouchEvent(TOUCH.END);

      expect(spyOnEnd).toHaveBeenCalledTimes(1);
    });
  });

  describe('mouse events', () => {
    const scene = new THREE.Scene();
    const camera = new THREE.PerspectiveCamera();

    it('should not invoke `onStart` if the `preventAction` function returned `true`', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );
      controls.preventAction = () => true;

      const spyOnStart = jest.spyOn(
        controls,
        // @ts-ignore
        'onStart',
      );

      fireMouseEvent(
        MOUSE.DOWN,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      expect(spyOnStart).not.toHaveBeenCalled();
    });

    it('should invoke `onStart` if the location is defined', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );
      const spyOnStart = jest.spyOn(
        controls,
        // @ts-ignore
        'onStart',
      );

      fireMouseEvent(
        MOUSE.DOWN,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      expect(spyOnStart).toHaveBeenCalledTimes(1);
    });

    it('should not invoke `onMove` if the `preventAction` function returned `true`', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );
      controls.preventAction = () => true;

      const spyOnMove = jest.spyOn(
        controls,
        // @ts-ignore
        'onMove',
      );

      fireMouseEvent(
        MOUSE.MOVE,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      expect(spyOnMove).not.toHaveBeenCalled();
    });

    it('should invoke `onMove` if the location is defined', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );
      const spyOnMove = jest.spyOn(
        controls,
        // @ts-ignore
        'onMove',
      );

      fireMouseEvent(
        MOUSE.MOVE,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      expect(spyOnMove).toHaveBeenCalledTimes(1);
    });

    it('should not invoke `onEnd` if the joystick is not attached', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );
      const spyOnEnd = jest.spyOn(
        controls,
        // @ts-ignore
        'onEnd',
      );

      controls.isJoystickAttached = false;

      fireMouseEvent(MOUSE.UP);

      expect(spyOnEnd).not.toHaveBeenCalled();
    });

    it('should invoke `onEnd` if the joystick is attached', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );
      const spyOnEnd = jest.spyOn(
        controls,
        // @ts-ignore
        'onEnd',
      );

      controls.isJoystickAttached = true;

      fireMouseEvent(MOUSE.UP);

      expect(spyOnEnd).toHaveBeenCalledTimes(1);
    });
  });

  describe('onStart', () => {
    const scene = new THREE.Scene();
    const camera = new THREE.PerspectiveCamera();

    it('should set the base anchor point to the clientX and clientY provided', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );

      fireMouseEvent(
        MOUSE.DOWN,
        {
          clientX: 0,
          clientY: 128,
        },
      );

      expect(controls.baseAnchorPoint).toEqual(new THREE.Vector2(0, 128));
    });

    it('should set interactionHasBegan to true', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );

      fireMouseEvent(
        MOUSE.DOWN,
        {
          clientX: 0,
          clientY: 128,
        },
      );

      expect(controls.interactionHasBegan).toEqual(true);
    });
  });

  describe('onMove', () => {
    const scene = new THREE.Scene();
    const camera = new THREE.PerspectiveCamera();

    it('should set the touch point to the clientX `0` and clientY `128`', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );

      fireTouchEvent(
        TOUCH.START,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      fireTouchEvent(
        TOUCH.MOVE,
        {
          clientX: 0,
          clientY: 128,
        },
      );

      expect(controls.touchPoint).toEqual(new THREE.Vector2(0, 128));
    });

    it('should invoke `attachJoystick` with the positionInScene mock vector', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );
      // @ts-ignore
      const spyAttachJoystick = jest.spyOn(controls, 'attachJoystick');

      fireTouchEvent(
        TOUCH.START,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      fireTouchEvent(
        TOUCH.MOVE,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      expect(spyAttachJoystick)
        .toHaveBeenCalledWith(new THREE.Vector2(100, 100));
    });

    it('should invoke `updateJoystickBallPosition` with the positionInScene', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );
      const spyUpdateJoystickBallPosition = jest
        // @ts-ignore
        .spyOn(controls, 'updateJoystickBallPosition');

      fireTouchEvent(
        TOUCH.START,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      fireTouchEvent(
        TOUCH.MOVE,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      fireTouchEvent(
        TOUCH.MOVE,
        {
          clientX: 0,
          clientY: 128,
        },
      );

      expect(spyUpdateJoystickBallPosition)
        .toHaveBeenCalledWith(0, 128, new THREE.Vector2(100, 100));
    });
  });

  describe('onEnd', () => {
    const scene = new THREE.Scene();
    const camera = new THREE.PerspectiveCamera();
    const spyGetObjectByName = jest.spyOn(scene, 'getObjectByName');

    it('should remove the joystick-base from the scene', () => {
      new JoystickControls(
        camera,
        scene,
      );

      fireTouchEvent(
        TOUCH.START,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      fireTouchEvent(
        TOUCH.MOVE,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      fireTouchEvent(
        TOUCH.END,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      expect(spyGetObjectByName).toHaveBeenCalledWith('joystick-base');
      expect(scene.getObjectByName('joystick-base')).toBeUndefined();
    });

    it('should remove the joystick-ball from the scene', () => {
      new JoystickControls(
        camera,
        scene,
      );

      fireTouchEvent(
        TOUCH.START,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      fireTouchEvent(
        TOUCH.MOVE,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      fireTouchEvent(
        TOUCH.END,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      expect(spyGetObjectByName).toHaveBeenLastCalledWith('joystick-ball');
      expect(scene.getObjectByName('joystick-ball')).toBeUndefined();
    });

    it('should set isJoystickAttached to false', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );
      controls.isJoystickAttached = true;
      controls.interactionHasBegan = true;

      fireTouchEvent(
        TOUCH.END,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      expect(controls.isJoystickAttached).toEqual(false);
    });

    it('should set interactionHasBegan to false', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );
      controls.isJoystickAttached = true;
      controls.interactionHasBegan = true;

      fireTouchEvent(
        TOUCH.END,
        {
          clientX: 0,
          clientY: 0,
        },
      );

      expect(controls.interactionHasBegan).toEqual(false);
    });
  });

  describe('attachJoystickUI', () => {
    const scene = new THREE.Scene();
    const camera = new THREE.PerspectiveCamera();

    it('should set the position to the mockPosition', () => {
      const controls = new JoystickControls(
        camera,
        scene,
      );
      const mockColor = 0x111111;
      const mockSize = 100;
      const mockPosition = new THREE.Vector3(1, 2, 3);
      const mockName = 'test-ui';
      // @ts-ignore
      controls.attachJoystickUI(
        mockName,
        mockPosition,
        mockColor,
        mockSize,
      );
      // @ts-ignore
      const sceneUIObject = scene.getObjectByName(mockName) as any;

      sceneUIObject.uuid = 'stripped-for-snapshot-test';
      sceneUIObject.geometry.uuid = 'stripped-for-snapshot-test';
      sceneUIObject.material.uuid = 'stripped-for-snapshot-test';
      sceneUIObject.parent.uuid = 'stripped-for-snapshot-test';

      expect(sceneUIObject).toMatchSnapshot();
    });
  });
});
