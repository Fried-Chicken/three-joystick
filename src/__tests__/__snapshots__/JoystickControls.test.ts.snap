// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoystickControls attachJoystickUI should set the position to the mockPosition 1`] = `
Object {
  "geometries": Array [
    Object {
      "radius": 100,
      "segments": 72,
      "thetaLength": 6.283185307179586,
      "thetaStart": 0,
      "type": "CircleGeometry",
      "uuid": "stripped-for-snapshot-test",
    },
  ],
  "materials": Array [
    Object {
      "color": 1118481,
      "colorWrite": true,
      "depthFunc": 3,
      "depthTest": false,
      "depthWrite": true,
      "emissive": 0,
      "opacity": 0.5,
      "reflectivity": 1,
      "refractionRatio": 0.98,
      "stencilFail": 7680,
      "stencilFunc": 519,
      "stencilFuncMask": 255,
      "stencilRef": 0,
      "stencilWrite": false,
      "stencilWriteMask": 255,
      "stencilZFail": 7680,
      "stencilZPass": 7680,
      "transparent": true,
      "type": "MeshLambertMaterial",
      "uuid": "stripped-for-snapshot-test",
    },
  ],
  "metadata": Object {
    "generator": "Object3D.toJSON",
    "type": "Object",
    "version": 4.5,
  },
  "object": Object {
    "geometry": "stripped-for-snapshot-test",
    "layers": 1,
    "material": "stripped-for-snapshot-test",
    "matrix": Array [
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
    ],
    "name": "test-ui",
    "renderOrder": 1,
    "type": "Mesh",
    "uuid": "stripped-for-snapshot-test",
  },
}
`;
